django/thunderstore/account/forms.py

```diff
@@ -0,0 +1,125 @@
+            user=user,
+            owner=self.cleaned_data["identity"],
+        )
+
+
+class DeleteServiceAccountForm(forms.Form):
+    def __init__(self, user: User, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.user = user
+        self.fields["service_account"] = forms.ModelChoiceField(
+            queryset=ServiceAccount.objects.filter(owner__members__user=user),
+        )
+
+    def clean_service_account(self) -> ServiceAccount:
+        service_account = self.cleaned_data["service_account"]
+        if not service_account.owner.can_delete_service_account(self.user):
+            raise ValidationError("Must be identity owner to delete a service account")
+        return service_account
+
+    def save(self) -> None:
+        self.cleaned_data["service_account"].delete()
+
+
+class EditServiceAccountForm(forms.Form):
+    nickname = forms.CharField(max_length=32)
+
+    def __init__(self, user: User, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.user = user
+        self.fields["service_account"] = forms.ModelChoiceField(
+            queryset=ServiceAccount.objects.filter(owner__members__user=user),
+        )
+
+    def clean_service_account(self) -> ServiceAccount:
+        service_account = self.cleaned_data["service_account"]
+        if not service_account.owner.can_edit_service_account(self.user):
+            raise ValidationError("Must be identity owner to edit a service account")
+        return service_account
+
+    def clean(self) -> None:
+        super().clean()
+        nickname = self.cleaned_data.get("nickname")
+        service_account = self.cleaned_data.get("service_account")
+
+        if not all((nickname, service_account)):
+            return
+
+    def save(self) -> ServiceAccount:
+        service_account = self.cleaned_data["service_account"]
+        service_account.user.first_name = self.cleaned_data["nickname"]
+        service_account.save()
+        return service_account
+
+
+class CreateTokenForm(forms.Form):
+    def __init__(self, user: User, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.user = user
+        self.fields["service_account"] = forms.ModelChoiceField(
+            queryset=ServiceAccount.objects.filter(owner__members__user=user),
+        )
+
+    def clean_service_account(self) -> ServiceAccount:
+        service_account = self.cleaned_data["service_account"]
+        if not service_account.owner.can_generate_service_account_token(self.user):
+            raise ValidationError(
+                "Must be identity owner to generate a service account token",
+            )
+        return service_account
+
+    @transaction.atomic
+    def save(self) -> Token:
+        service_account_user = self.cleaned_data["service_account"].user
+        Token.objects.filter(user=service_account_user).delete()
+        return Token.objects.create(user=service_account_user)
```

Comment:
I feel like there's a better way of doing this

Reply:
Looks good enough tome